### PR TITLE
AP_GPS: improve GPS switching logic to reduce switch frequency

### PR DESCRIPTION
**Pull request by magicrub**
param "GPS_AUTO_SWITCH" now has a 3rd option
- 0:Disabled,1:Use the better lock,2:Treat GPS2 as backup

follow along here: https://github.com/diydrones/ardupilot/issues/3754

### CODE REVIEW

**Comment by magicrub** [comment]
per a mumble chat between me and @WickedShell , I re-wrote this section to be without a for-loop and hard-coded the indexes.

File: libraries/AP_GPS/AP_GPS.cpp [hunk 1]
```diff
@@ -430,12 +430,12 @@ AP_GPS::update(void)
             break;
 
         case GPS_AUTO_SWITCH_BETTER_LOCK:
-            if (state[0].status > state[1].status) {
-                // there is a higher status lock, change GPS
-                primary_instance = 0;
-            } else if (state[1].status > state[0].status) {
-                // there is a higher status lock, change GPS
-                primary_instance = 1;
+            if (state[0].status > state[1].status && (state[0].num_sats >= state[1].status + 2)) {
```
Comment (OXINARF) [inline libraries/AP_GPS/AP_GPS.cpp]: This should be `num_stats + 2` right?

File: libraries/AP_GPS/AP_GPS.cpp [hunk 2]
```diff
@@ -430,12 +430,12 @@ AP_GPS::update(void)
...
+                // there is a higher status lock or 2+ sats, change GPS
+                primary_instance_change_request = 0;
+            } else if (state[1].status > state[0].status && (state[1].num_sats >= state[0].status + 2)) {
```
Comment (OXINARF) [inline libraries/AP_GPS/AP_GPS.cpp]: And here too

File: libraries/AP_GPS/AP_GPS.cpp (hunk 1 above)
Comment (magicrub) [inline libraries/AP_GPS/AP_GPS.cpp]: ahh yes, good catch!

**Comment by magicrub** [comment]
Turns out we have a better solution in mind, closing this. Will get replaced later

**Pull request closed without merging**